journalbeat/reader/journal.go

```diff
@@ -188,40 +182,31 @@ func (r *Reader) seek(cursor string) {
 // Next waits until a new event shows up and returns it.
 // It blocks until an event is returned or an error occurs.
 func (r *Reader) Next() (*beat.Event, error) {
-	event, err := r.readEvent()
-	if err != nil {
-		return nil, err
-	}
-	if event != nil {
-		return event, nil
-	}
-
-	return r.waitUntilNewEventOrError()
-}
-
-func (r *Reader) waitUntilNewEventOrError() (*beat.Event, error) {
 	for {
 		select {
 		case <-r.done:
 			return nil, nil
-		case c := <-r.changesChan:
+		default:
+		}
+
+		c, err := r.journal.Next()
+		if err != nil && err != io.EOF {
+			return nil, err
+		}
+
+		// error while reading next entry
+		if c < 0 {
```

Comment:
according to docs c is uint64. Why does this compile?

Reply:
The parameter of `sd_journal_wait` is `uint64`. Its return value is `int`.
* https://godoc.org/github.com/coreos/go-systemd/sdjournal#Journal.Wait
* https://manpages.debian.org/testing/libsystemd-dev/sd_journal_wait.3.en.html#RETURN_VALUE

Reply:
This `c` is not from the call to `Wait`, but from `Next`. You even shadow/redefine `c` when calling `Wait`.
Sometimes it's a good idea not to shadow/reuse variable names.

See: https://godoc.org/github.com/coreos/go-systemd/sdjournal#Journal.Next

Seems like sdjournal API is somewhat inconsistent here.

How about wrapping `Wait`, so to reduce the chance of confusion (and reducing level of nesting by 1):

```
func journalWait(j *Journal, dur time.Duration) (has bool, err error) {
  switch c := j.Wait(dur) {
  case sdjournal.SD_JOURNAL_NOP:
    return false, nil
  case sdjournal.SD_JOURNAL_APPEND, sdjournal.SD_JOURNAL_INVALIDATE:
    return true, nil
  default:
    if c < 0 {
 	  return false, fmt.Errorf("error while waiting for event: %+v", syscall.Errno(-c))
    }
    return false, fmt.Errorf("Unknown return code from Wait: %d\n", c)
  }
}
```

This can be used like:

```
for {
  ...

  c, err := r.journal.Next()
  if err != nil && err != io.EOF {
    return nil, err
  }
  if c == 0 {
    has, err := journalWait(r.journal, 100 * time.Millisecond)
    if err != nil {
      return nil, err
    }

    if !has {
      continue
    }
  }

  ...
}
```

Interestingly all other calls on journal properly return error values, but Wait does not. Was `Wait` added by the original developers or some contributor?
Maybe we can open a PR/issue to enhance consistency in the sdjournal API. And add support for getting the FD itself. Plus expose some other missing functions like sd_journal_process & co.

Reply:
TIL go compiler does not warn/error on not-satisfiable comparisons on uint types: https://play.golang.org/p/g0NFrotaKd-

Reply:
Question: What shall we do if Wait returns an unknown value >0?
1. treat it as error (my sample code) ?
2. treat it like SD_JOURNAL_APPEND (current implementation) ?
3. treat it like SD_JOURNAL_NOP ?

Reply:
In my implementation (or at least what I intended to implement before messing up with copy-pasting my previous code... :D), an error is logged and nil is returned from `Next`. Thus, the reader tries to read a new entry again.

Reply:
`Wait` seems like it was added later, when `sd-journal.h` became a dependency. 
I would like to open a PR there to add support for the functions we need to check for events without using `Wait`. (As you proposed it in a previous PR.)

Reply:
This conditional `c < 0` is still a no-op, as it will always return `false`. Let's remove it.

Reply:
Done.